infra/event-sync/src/main/java/com/simprints/infra/eventsync/sync/up/tasks/EventUpSyncTask.kt

```diff
@@ -74,115 +78,104 @@ internal class EventUpSyncTask @Inject constructor(
         Simber.d("[EVENT_REPO] Uploading")
-
-        eventRepository.getAllClosedSessionIds(projectId).forEach { sessionId ->
-            // The events will include the SessionCaptureEvent event
-            Simber.d("[EVENT_REPO] Uploading session $sessionId")
-            try {
-                eventRepository.getEventsFromSession(sessionId).let {
-                    attemptEventUpload(
-                        it, projectId,
-                        canSyncAllDataToSimprints,
-                        canSyncBiometricDataToSimprints,
-                        canSyncAnalyticsDataToSimprints
-                    )
-                    this.emit(it.size)
-                }
-            } catch (ex: Exception) {
-                if (ex is JsonParseException || ex is JsonMappingException) {
-                    attemptInvalidEventUpload(projectId, sessionId)?.let { this.emit(it) }
-                } else {
-                    Simber.i("Failed to un-marshal events for $sessionId")
+        try {
+            val sessionScopes = eventRepository.getAllClosedSessions(projectId).associateWith {
+                try {
+                    eventRepository.getEventsFromSession(it.id)
+                        .also { listOfEvents -> emit(listOfEvents.size) }
+                } catch (ex: JsonParseException) {
+                    Simber.i("Failed to un-marshal events")
                     Simber.e(ex)
+                    null
                 }
             }
-        }
-    }
 
-    private suspend fun attemptEventUpload(
-        events: List<Event>, projectId: String,
-        canSyncAllDataToSimprints: Boolean,
-        canSyncBiometricDataToSimprints: Boolean,
-        canSyncAnalyticsDataToSimprints: Boolean
-    ) {
-        try {
-            val filteredEvents = when {
-                canSyncAllDataToSimprints -> {
-                    events
-                }
-                canSyncBiometricDataToSimprints -> {
-                    events.filter {
-                        it is EnrolmentEventV2 || it is PersonCreationEvent || it is FingerprintCaptureBiometricsEvent || it is FaceCaptureBiometricsEvent
-                    }
+            val corruptedScopes = sessionScopes.filterValues { it == null }.keys
+
+            // Re-emitting the number of uploaded corrupted events
+            attemptInvalidEventUpload(projectId, corruptedScopes).collect { emit(it) }
+
+            val scopesToUpload = sessionScopes
+                .filterValues { it != null }
+                .mapValues { (_, events) ->
+                    events?.let { filterEventsToUpSync(events, config) }.orEmpty()
                 }
-                canSyncAnalyticsDataToSimprints -> {
-                    events.filterNot {
-                        it is FingerprintCaptureBiometricsEvent || it is FaceCaptureBiometricsEvent
-                    }
+            if (scopesToUpload.isNotEmpty()) {
+                eventRemoteDataSource.post(projectId, scopesToUpload)
```

Comment:
We are moving away from 1 session per request. For now all are in a single request, eventually, it will be batched.

Reply:
Will this not cause issues if the devices have bad connections and a lot of sessions happen as to succeed in uploading one session, they will need to have enough connection to upload all their sessions from their database. 
I think this is why in the current implementation we are doing 1 session per request

Reply:
@luhmirin-s - why are we moving away from it?
Same as Yoann, I'm worried about devices that have been offline for a long time and have accumulated 100s or even 1000s of sessions. There might be issues with the request itself being harder to upload successfully (especially in bad connectivity) - leading to more unsuccessful attempts. And at extremes it might simply become impossible to upload - due to OOM crashes or other issues with a huge single request.

Reply:
@ybourgery @BurningAXE Batched session upload is one of the optimisations proposed in https://docs.google.com/document/d/1CxnIqrtI8UcqYDDtYyJS0XfCcvGm42gxbsJBm6lHdzM/edit#heading=h.p5w7n6lxzjw

In short - uploading a single session per request creates huge overhead and batching them would allow us to improve data/overhead ratio and increase upload speed. In newer PR session upload has been changed to arbitrary batch size for some baseline, eventually there is an idea to allow batch sizes to be configured on BE side to find the optimal value. 

Anyhow, I suggest moving this discussion from already closed PR to either the design doc or slack.